src/cargo/core/resolver/mod.rs

```diff
@@ -1255,10 +1265,41 @@ fn activate_deps_loop(
                             .flat_map(|x| x)
                             .find(|con| cx.is_conflicting(None, con))
                         {
-                            conflicting_activations.extend(conflicting.clone());
+                            let mut conflicting = conflicting.clone();
+                            conflicting.remove(&pid);
+
+                            conflicting_activations.extend(conflicting);
                             has_past_conflicting_dep = true;
                         }
                     }
+                    if !has_past_conflicting_dep {
+                        // TODO: this is ugly, replace!
```

Comment:
Ok I'm gonna do my best to rephrase what's happening here in english and you can tell me how wrong I am!

Here we're iterating over *all* remaining work we have to do (aka all of `remaining_deps`). We're looking at every single remaining `Dependency` left to work with. One of those `Dependency` structures may be known to fail because of a known set of conflicting activated packages. If that conflicting set includes the candidate we just activated *and* we've activated everything in the set then we know this `Dependency` is going to fail.

Since we know this `Dependency` is going to fail we want to extend our known set of conflicting activations. These conflicting activations are for when our own candidate can't be activated, and we've just deduced that activating ourselves will cause something else to fail.

After this though... I think I get a little lost. We take this conflicting set for some remote dependency, we remove ourselves from it, and we insert that remote dependency's parent with the reason we just removed. How come we do the swich here?

In any case we finally reach `conflicting_activations.extend(conflict)` after all that. If we entirely fail activation then this ensures that `conflicting_activations` contains an exhaustive set as to the reasons why we failed to activated.

I think this all makes sense? It seems plausible at least!

Reply:
Exactly correct! The switch is just a kinda change in perspective. 

The conflict we found is "**other dep** will not succeed if we are activated." We want to add "our dep will not succeed if **other dep** is in `remaining_deps`" but that is not how the cache is set up. So we add the less general but much faster, "our dep will not succeed if **other dep**'s parent is activated". 

We can remove ourselves from the list because we know that if our dep is being looked at then we are going to be activated. We have to remove ourselves from the list as when our parent considers if we will succeed we won't have been activated yet.

> It seems plausible at least!

This resolver things is complicated enough, and I have been wrong about things I was sure of enough, that we should not merge unless we are sure it is correct and the comments make it possible to pick it up agen.

Reply:
Ok I think that definitely makes sense. I think it's a direct implication that if the other dep's parent is activated that means that other dep will eventually be in `remaining_deps`, which means that if we're activated again it'll cause that to fail.

That definitely cleared things up for me! I'm definitely much more confident in this now